kernel/src/table_features/column_mapping.rs

```diff
@@ -25,18 +29,135 @@ pub(crate) fn column_mapping_mode(
+                )));
+            }
+            (ColumnMappingMode::None, Some(_)) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is not enabled but field '{annotation}' is annotated with {}",
+                    column_name()
+                )));
+            }
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, None) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is enabled but field '{}' lacks the {annotation} annotation",
+                    column_name()
+                )));
+            }
+        }
+
+        let annotation = "delta.columnMapping.id";
+        match (self.mode, field.metadata.get(annotation)) {
+            // Both Id and Name modes require a field ID annotation; None mode forbids it.
+            (ColumnMappingMode::None, None) => {}
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, Some(MetadataValue::Number(_))) => {}
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, Some(_)) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "The {annotation} annotation on field '{}' must be a number",
+                    column_name()
+                )));
+            }
+            (ColumnMappingMode::None, Some(_)) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is not enabled but field '{}' is annotated with {annotation}",
+                    column_name()
+                )));
+            }
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, None) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is enabled but field '{}' lacks the {annotation} annotation",
+                    column_name()
+                )));
+            }
+        }
+    }
+}
+
+impl<'a> SchemaTransform<'a> for ValidateColumnMappings<'a> {
+    // Override array element and map key/value for better error messages
+    fn transform_array_element(&mut self, etype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform_inner_type(etype, "<array element>")
+    }
+    fn transform_map_key(&mut self, ktype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform_inner_type(ktype, "<map key>")
+    }
+    fn transform_map_value(&mut self, vtype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform_inner_type(vtype, "<map value>")
+    }
+    fn transform_struct_field(&mut self, field: &'a StructField) -> Option<Cow<'a, StructField>> {
+        if self.err.is_none() {
+            self.path.push(&field.name);
+            self.check_annotations(field);
+            let _ = self.recurse_into_struct_field(field);
+            self.path.pop();
+        }
+        None
+    }
+}
+
 #[cfg(test)]
 mod tests {
```

Comment:
I would like to see some of the validator error cases exercised. Either in this PR or a followup one :)

Reply:
Done!